pkg/autoscaler/aggregation/bucketing2.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2019 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aggregation
+
+import (
+	"sync"
+	"time"
+)
+
+// TimedFloat64Buckets2 keeps buckets that have been collected at a certain time.
+type TimedFloat64Buckets2 struct {
+	bucketsMutex sync.RWMutex
+	buckets      []float64Bucket
+	lastWrite    time.Time
+
+	granularity time.Duration
+	window      time.Duration
+}
+
+// NewTimedFloat64Buckets2 generates a new TimedFloat64Buckets2 with the given
+// granularity.
+func NewTimedFloat64Buckets2(window, granularity time.Duration) *TimedFloat64Buckets2 {
+	// Number of buckets is window divided by granurality.
+	// e.g. 60s / 2s = 30.
+	nb := int(window / granularity)
+	return &TimedFloat64Buckets2{
+		buckets:     make([]float64Bucket, nb),
+		granularity: granularity,
+		window:      window,
+	}
+}
+
+// Record adds a value with an associated time to the correct bucket.
+func (t *TimedFloat64Buckets2) Record(now time.Time, name string, value float64) {
+	bucketTime := now.Truncate(t.granularity)
+	// I don't think this run in 2038 :-)
+	bucketIdx := int(bucketTime.Unix()) % len(t.buckets)
+
+	t.bucketsMutex.Lock()
+	defer t.bucketsMutex.Unlock()
+
+	if t.lastWrite != bucketTime {
+		// Reset the value.
+		t.buckets[bucketIdx] = float64Bucket{}
+		// Update the last write time.
+		t.lastWrite = bucketTime
+	}
+	t.buckets[bucketIdx].record(name, value)
+}
+
+// isEmpty returns whether or not there are no values currently stored.
+// isEmpty requires t.bucketMux to be held.
+func (t *TimedFloat64Buckets2) isEmpty(now time.Time) bool {
+	now = now.Truncate(t.granularity)
+	return now.Sub(t.lastWrite) >= t.window
+}
+
+// ForEachBucket calls the given Accumulator function for each bucket.
+// Returns true if any data was recorded.
+func (t *TimedFloat64Buckets2) ForEachBucket(now time.Time, accs ...Accumulator) bool {
+	now = now.Truncate(t.granularity)
+	t.bucketsMutex.RLock()
+	defer t.bucketsMutex.RUnlock()
+
+	if t.isEmpty(now) {
+		return false
+	}
+
+	// So number of buckets we can process is len(buckets)-(now-lastWrite)/granularity.
+	// Since isEmpty returned false, we know this is at least 1 bucket.
+	nb := len(t.buckets) - int(now.Sub(t.lastWrite)/t.granularity)
+	tb := t.lastWrite // Always aligned with granularity.
+	for i := 0; i < nb; i++ {
+		ti := int(tb.Unix()) % len(t.buckets)
```

Comment:
It feels like we only need to compute this once and then walk the underlying array via its index rather than computing the correct timestamp `now - granularity`.

Reply:
we need time still to call in `acc`.

Reply:
Though I changed to use index for actual access.